File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/polynomial/DefaultPolynomialExtrapolation.java
package org.kevoree.modeling.api.polynomial;

/**
 * Created by assaa_000 on 28/10/2014.
 */

import org.kevoree.modeling.api.polynomial.util.DataSample;
import org.kevoree.modeling.api.polynomial.util.PolynomialFitEjml;
import org.kevoree.modeling.api.polynomial.util.Prioritization;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by assaa_000 on 23/10/2014.
 */
public class DefaultPolynomialExtrapolation implements PolynomialExtrapolation {

    private double[] weights;
    private Long timeOrigin;
    private List<DataSample> samples = new ArrayList<DataSample>();
    private int degradeFactor;
    private Prioritization prioritization;
    private int maxDegree;
    private double toleratedError;
    //TODO add to save
    private long lastIndex = -1;

    public DefaultPolynomialExtrapolation(long timeOrigin, double toleratedError, int maxDegree, int degradeFactor, Prioritization prioritization) {
        this.timeOrigin = timeOrigin;
        this.degradeFactor = degradeFactor;
        this.prioritization = prioritization;
        this.maxDegree = maxDegree;
        this.toleratedError = toleratedError;
    }

    public List<DataSample> getSamples() {
        return samples;
    }

    public int getDegree() {
        if (weights == null) {
            return -1;
        } else {
            return weights.length - 1;
        }
    }

    public Long getTimeOrigin() {
        return timeOrigin;
    }

    private double getMaxErr(int degree, double toleratedError, int maxDegree, Prioritization prioritization) {
        double tol = toleratedError;
        if (prioritization == Prioritization.HIGHDEGREES) {
            tol = toleratedError / Math.pow(2, maxDegree - degree);
        } else if (prioritization == Prioritization.LOWDEGREES) {
            tol = toleratedError / Math.pow(2, degree + 0.5);
        } else if (prioritization == Prioritization.SAMEPRIORITY) {
            tol = toleratedError * degree * 2 / (2 * maxDegree);
        }
        return tol;
    }


    private void internal_feed(Long time, double value) {
        //If this is the first point in the set, add it and return
        if (weights == null) {
            weights = new double[1];
            weights[0] = value;
            timeOrigin = time;
            samples.add(new DataSample(time, value));
        }
    }

    private double maxError(double[] computedWeights, long time, double value) {
        double maxErr = 0;
        double temp = 0;
        DataSample ds;
        for (int i = 0; i < samples.size(); i++) {
            ds = samples.get(i);
            double val = internal_extrapolate(ds.time, computedWeights);
            temp = Math.abs(val - ds.value);
            if (temp > maxErr) {
                maxErr = temp;
            }
        }
        temp = Math.abs(internal_extrapolate(time, computedWeights) - value);
        if (temp > maxErr) {
            maxErr = temp;
        }
        return maxErr;
    }

    public boolean comparePolynome(DefaultPolynomialExtrapolation p2, double err) {
        if (weights.length != p2.weights.length) {
            return false;
        }
        for (int i = 0; i < weights.length; i++) {
            if (Math.abs(weights[i] - weights[i]) > err) {
                return false;
            }
        }
        return true;
    }

    private double internal_extrapolate(long time, double[] weights) {
        double result = 0;
        double t = ((double) (time - timeOrigin)) / degradeFactor;
        double power = 1;
        for (int j = 0; j < weights.length; j++) {
            result += weights[j] * power;
            power = power * t;
        }
        return result;
    }

    @Override
    public double extrapolate(long time) {
        return internal_extrapolate(time, weights);
    }


    @Override
    public boolean insert(long time, double value) {
        //If this is the first point in the set, add it and return
        if (weights == null) {
            internal_feed(time, value);
            lastIndex = time;
            return true;
        }
        double maxError = getMaxErr(this.getDegree(), toleratedError, maxDegree, prioritization);
        //If the current model fits well the new value, return
        if (Math.abs(extrapolate(time) - value) <= maxError) {
            samples.add(new DataSample(time, value));
            lastIndex = time;
            return true;
        }
        //If not, first check if we can increase the degree
        int deg = getDegree();
        if (deg < maxDegree) {
            deg++;
            int ss = Math.min(deg * 2, samples.size());
            double[] times = new double[ss + 1];
            double[] values = new double[ss + 1];
            int current = samples.size();
            for (int i = 0; i < ss; i++) {
                DataSample ds = samples.get(i * current / ss);
                times[i] = ((double) (ds.time - timeOrigin)) / degradeFactor;
                values[i] = ds.value;
            }
            times[ss] = ((double) (time - timeOrigin)) / degradeFactor;
            values[ss] = value;
            PolynomialFitEjml pf = new PolynomialFitEjml(deg);
            pf.fit(times, values);
            if (maxError(pf.getCoef(), time, value) <= maxError) {
                weights = new double[pf.getCoef().length];
                for (int i = 0; i < pf.getCoef().length; i++) {
                    weights[i] = pf.getCoef()[i];
                }
                samples.add(new DataSample(time, value));
                lastIndex = time;
                return true;
            }
        }
        return false;
    }

    @Override
    public long lastIndex() {
        return 0;
    }

    private static final char sep = '|';

    @Override
    public String save() {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < weights.length; i++) {
            if (i != 0) {
                builder.append(sep);
            }
            builder.append(weights[i]);
        }
        return builder.toString();
    }

    @Override
    public void load(String payload) {
        String[] elems = payload.split(sep + "");
        weights = new double[elems.length];
        for (int i = 0; i < elems.length; i++) {
            weights[i] = Long.parseLong(elems[i]);
        }
    }

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/polynomial/DefaultPolynomialExtrapolation.ts
///<reference path="util/DataSample.ts"/>
///<reference path="util/PolynomialFitEjml.ts"/>
///<reference path="util/Prioritization.ts"/>
///<reference path="../../../../../java/util/JUArrayList.ts"/>
///<reference path="../../../../../java/util/JUList.ts"/>

class DefaultPolynomialExtrapolation implements PolynomialExtrapolation {

  private weights: number[] = null;
  private timeOrigin: number = null;
  private samples: JUList<DataSample> = new JUArrayList<DataSample>();
  private degradeFactor: number = 0;
  private prioritization: Prioritization = null;
  private maxDegree: number = 0;
  private toleratedError: number = 0;
  private lastIndex: number = -1;
  private static sep: string = '|';

  constructor(timeOrigin: number, toleratedError: number, maxDegree: number, degradeFactor: number, prioritization: Prioritization) {
    this.timeOrigin = timeOrigin;
    this.degradeFactor = degradeFactor;
    this.prioritization = prioritization;
    this.maxDegree = maxDegree;
    this.toleratedError = toleratedError;
  }

  public getSamples(): JUList<DataSample> {
    return this.samples;
  }

  public getDegree(): number {
    if (this.weights == null) {
      return -1;
    } else {
      return this.weights.length - 1;
    }
  }

  public getTimeOrigin(): number {
    return this.timeOrigin;
  }

  private getMaxErr(degree: number, toleratedError: number, maxDegree: number, prioritization: Prioritization): number {
    var tol: number = toleratedError;
    if (prioritization == Prioritization.HIGHDEGREES) {
      tol = toleratedError / Math.pow(2, maxDegree - degree);
    } else {
      if (prioritization == Prioritization.LOWDEGREES) {
        tol = toleratedError / Math.pow(2, degree + 0.5);
      } else {
        if (prioritization == Prioritization.SAMEPRIORITY) {
          tol = toleratedError * degree * 2 / (2 * maxDegree);
        }
      }
    }
    return tol;
  }

  private internal_feed(time: number, value: number): void {
    if (this.weights == null) {
      this.weights = new Array();
      this.weights[0] = value;
      this.timeOrigin = time;
      this.samples.add(new DataSample(time, value));
    }
  }

  private maxError(computedWeights: number[], time: number, value: number): number {
    var maxErr: number = 0;
    var temp: number = 0;
    var ds: DataSample;
    for (var i: number = 0; i < this.samples.size(); i++) {
      ds = this.samples.get(i);
      var val: number = this.internal_extrapolate(ds.time, computedWeights);
      temp = Math.abs(val - ds.value);
      if (temp > maxErr) {
        maxErr = temp;
      }
    }
    temp = Math.abs(this.internal_extrapolate(time, computedWeights) - value);
    if (temp > maxErr) {
      maxErr = temp;
    }
    return maxErr;
  }

  public comparePolynome(p2: DefaultPolynomialExtrapolation, err: number): boolean {
    if (this.weights.length != p2.weights.length) {
      return false;
    }
    for (var i: number = 0; i < this.weights.length; i++) {
      if (Math.abs(this.weights[i] - this.weights[i]) > err) {
        return false;
      }
    }
    return true;
  }

  private internal_extrapolate(time: number, weights: number[]): number {
    var result: number = 0;
    var t: number = (<number>(time - this.timeOrigin)) / this.degradeFactor;
    var power: number = 1;
    for (var j: number = 0; j < weights.length; j++) {
      result += weights[j] * power;
      power = power * t;
    }
    return result;
  }

  public extrapolate(time: number): number {
    return this.internal_extrapolate(time, this.weights);
  }

  public insert(time: number, value: number): boolean {
    if (this.weights == null) {
      this.internal_feed(time, value);
      this.lastIndex = time;
      return true;
    }
    var maxError: number = this.getMaxErr(this.getDegree(), this.toleratedError, this.maxDegree, this.prioritization);
    if (Math.abs(this.extrapolate(time) - value) <= maxError) {
      this.samples.add(new DataSample(time, value));
      this.lastIndex = time;
      return true;
    }
    var deg: number = this.getDegree();
    if (deg < this.maxDegree) {
      deg++;
      var ss: number = Math.min(deg * 2, this.samples.size());
      var times: number[] = new Array();
      var values: number[] = new Array();
      var current: number = this.samples.size();
      for (var i: number = 0; i < ss; i++) {
        var ds: DataSample = this.samples.get(i * current / ss);
        times[i] = (<number>(ds.time - this.timeOrigin)) / this.degradeFactor;
        values[i] = ds.value;
      }
      times[ss] = (<number>(time - this.timeOrigin)) / this.degradeFactor;
      values[ss] = value;
      var pf: PolynomialFitEjml = new PolynomialFitEjml(deg);
      pf.fit(times, values);
      if (this.maxError(pf.getCoef(), time, value) <= maxError) {
        this.weights = new Array();
        for (var i: number = 0; i < pf.getCoef().length; i++) {
          this.weights[i] = pf.getCoef()[i];
        }
        this.samples.add(new DataSample(time, value));
        this.lastIndex = time;
        return true;
      }
    }
    return false;
  }

  public lastIndex(): number {
    return 0;
  }

  public save(): string {
    var builder: StringBuilder = new StringBuilder();
    for (var i: number = 0; i < this.weights.length; i++) {
      if (i != 0) {
        builder.append(DefaultPolynomialExtrapolation.sep);
      }
      builder.append(this.weights[i]);
    }
    return builder.toString();
  }

  public load(payload: string): void {
    var elems: string[] = payload.split(DefaultPolynomialExtrapolation.sep + "");
    this.weights = new Array();
    for (var i: number = 0; i < elems.length; i++) {
      this.weights[i] = Long.parseLong(elems[i]);
    }
  }

}

